server/src/main/java/org/elasticsearch/health/node/LocalHealthMonitor.java

```diff
@@ -91,63 +116,226 @@ private void registerListeners() {
+     */
+    static class HealthMonitoringTask implements Runnable, Scheduler.Cancellable {
+
+        private final TimeValue interval;
+        private final String executor;
+        private final Scheduler scheduler;
+        private final ClusterService clusterService;
+        private final DiskCheck diskCheck;
+        private final Client client;
+
+        private volatile boolean run = true;
+        private final AtomicReference<DiskHealthInfo> lastReportedDiskHealthInfo;
+        private final AtomicReference<String> lastSeenHealthNode;
+
+        HealthMonitoringTask(
+            TimeValue interval,
+            String executor,
+            Scheduler scheduler,
+            AtomicReference<DiskHealthInfo> lastReportedDiskHealthInfo,
+            AtomicReference<String> lastSeenHealthNode,
+            DiskCheck diskCheck,
+            ClusterService clusterService,
+            Client client
+        ) {
+            this.interval = interval;
+            this.executor = executor;
+            this.scheduler = scheduler;
+            this.lastReportedDiskHealthInfo = lastReportedDiskHealthInfo;
+            this.lastSeenHealthNode = lastSeenHealthNode;
+            this.clusterService = clusterService;
+            this.diskCheck = diskCheck;
+            this.client = client;
+            scheduler.schedule(this, TimeValue.ZERO, executor);
+        }
+
+        @Override
+        public boolean cancel() {
+            final boolean result = run;
+            run = false;
+            return result;
+        }
+
+        @Override
+        public boolean isCancelled() {
+            return run == false;
+        }
+
+        /**
+         * This method evaluates the health info of this node and if there is a change it sends an update request to the health node
+         */
+        @Override
+        public void run() {
+            boolean nextRunScheduled = false;
+            Runnable runOnceScheduleNextRunIfNecessary = new RunOnce(this::scheduleNextRunIfNecessary);
+            try {
+                ClusterState clusterState = clusterService.state();
+                HealthMetadata healthMetadata = HealthMetadata.getFromClusterState(clusterState);
+                if (healthMetadata != null) {
+                    DiskHealthInfo previousHealth = this.lastReportedDiskHealthInfo.get();
+                    DiskHealthInfo currentHealth = diskCheck.getHealth(healthMetadata, clusterState);
+                    if (currentHealth.equals(previousHealth) == false) {
+                        String nodeId = clusterService.localNode().getId();
+                        String healthNodeId = lastSeenHealthNode.get();
+                        ActionListener<AcknowledgedResponse> listener = ActionListener.wrap(response -> {
+                            // Update the last reported value only if the health node hasn't changed.
+                            if (Objects.equals(healthNodeId, lastSeenHealthNode.get())
```

Comment:
We probably ought to check the value of `run` here at the beginning of the `if` block right? Since there could theoretically be a lot of time between when `client.execute()` was called and when this listener is called (it's possible that in the meantime this run could have been cancelled and there could have been other requests made I think).

Reply:
Yes indeed. Good catch!

Reply:
Hm, interesting thought. So you suggest to check if it was cancelled for two reasons:
- _Because it might have changed at a different thread_
In this case, I think we are covered since we only update these fields if `lastReportedDiskHealthInfo` hasn't changed and the health node is the same. Right?

- _Because, the monitoring task was cancelled._
I completely agree with you that we should check this at the beginning, but I wouldn't want to check it in multiple places in the code because then there will be arbitrary places that the execution will behave like it was interrupted if it was cancelled, I feel this would make it more difficult to argue about the code. Furthermore, I don't think that it's bad that these fields get updated after the task was cancelled since the execution was already in progress and it wasn't cancelled. In the end, the health info was sent to the current health node, and since this is what the current health node just received we should update them. What do you think?

Reply:
The reason for the 2nd check was b/c it's potentially running much later since it's a listener. If `previousHealth` were null for both this one would still incorrectly get set. Maybe as you say though that doesn't matter since it came from the correct health node? It's probably the same result.